File: components/signin/public/identity_manager/android/java/src/org/chromium/components/signin/identitymanager/CoreAccountId.java
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.signin.identitymanager;

import android.support.annotation.NonNull;

/**
 * A wrapper around Gaia ID that represents a stable account identifier.
 *
 * This wrapper helps to make sure that code using accounts doesn't accidentally use account name in
 * place of Gaia ID and vice versa.
 *
 * This class has a native counterpart called CoreAccountId.
 */
public class CoreAccountId {
    private final String mGaiaId;

    /**
     * Constructs a new CoreAccountId from a String representation of Gaia ID.
     */
    public CoreAccountId(@NonNull String gaiaId) {
        assert gaiaId != null;
        // Check that a user email is not used by mistake.
        assert !gaiaId.contains("@");

        mGaiaId = gaiaId;
    }

    public String getGaiaIdAsString() {
        return mGaiaId;
    }

    @Override
    public String toString() {
        return mGaiaId;
    }

    @Override
    public int hashCode() {
        return mGaiaId.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof CoreAccountId)) return false;
        CoreAccountId other = (CoreAccountId) obj;
        return mGaiaId.equals(other.mGaiaId);
    }
}

File: src/google_apis/gaia/core_account_id.cc
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "google_apis/gaia/core_account_id.h"

CoreAccountId::CoreAccountId() = default;

CoreAccountId::~CoreAccountId() = default;

CoreAccountId::CoreAccountId(const char* id) : id(id) {}

CoreAccountId::CoreAccountId(std::string&& id) : id(std::move(id)) {}

CoreAccountId::CoreAccountId(const std::string& id) : id(id) {}

CoreAccountId::operator std::string() const {
  return id;
}

bool CoreAccountId::empty() const {
  return id.empty();
}

bool operator<(const CoreAccountId& lhs, const CoreAccountId& rhs) {
  return lhs.id < rhs.id;
}

bool operator==(const CoreAccountId& lhs, const CoreAccountId& rhs) {
  return lhs.id == rhs.id;
}

bool operator!=(const CoreAccountId& lhs, const CoreAccountId& rhs) {
  return lhs.id != rhs.id;
}

std::ostream& operator<<(std::ostream& out, const CoreAccountId& a) {
  return out << a.id;
}
